File: Objective-C/Common/PaymentTransactionDetails.m
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
A BaseViewController subclass that provides details about a purchase. The purchase contains the product identifier, transaction identifier, and
 transaction date for a regular purchase. It includes the content identifier, content version, and content length for a hosted product. It
 contains the original transaction's identifier and date for a restored purchase.
*/

@import StoreKit;
#import "Section.h"
#import "AppConfiguration.h"
#import "PaymentTransactionDetails.h"

typedef NS_ENUM(NSInteger, PaymentTransactionDetailsSection) {
	PaymentTransactionDetailsSectionInformation = 0,
	PaymentTransactionDetailsSectionDownload,
	PaymentTransactionDetailsSectionOriginalTransaction
};

NSString *const kBasicCellIdentifier = @"basic";
NSString *const kCustomCellIdentifier = @"custom";

@implementation PaymentTransactionDetails
#pragma mark - UITable​View​Data​Source

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
	Section *section = (self.data)[indexPath.section];

	if ([section.name isEqualToString:PCSPurchasesDownload] || [section.name isEqualToString:PCSPaymentTransactionDetailsOriginalTransaction]) {
		return [tableView dequeueReusableCellWithIdentifier:kCustomCellIdentifier forIndexPath:indexPath];
	} else {
		return [tableView dequeueReusableCellWithIdentifier:kBasicCellIdentifier forIndexPath:indexPath];
	}
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
	Section *section = (self.data)[indexPath.section];
	NSArray *transactions = section.elements;

	if ([section.name isEqualToString:PCSPurchasesDownload]) {
		NSDictionary *dictionary = transactions[indexPath.row];

		switch (indexPath.row) {
			case PaymentTransactionDetailsSectionInformation:
				cell.textLabel.text = PCSPaymentTransactionDetailsLabelsIdentifier;
				cell.detailTextLabel.text = dictionary[PCSPaymentTransactionDetailsLabelsIdentifier];
				break;
			case PaymentTransactionDetailsSectionDownload:
				cell.textLabel.text = PCSPaymentTransactionDetailsLabelsContentVersion;
				cell.detailTextLabel.text = dictionary[PCSPaymentTransactionDetailsLabelsContentVersion];
				break;
			case PaymentTransactionDetailsSectionOriginalTransaction:
				cell.textLabel.text = PCSPaymentTransactionDetailsLabelsContentLength;
				cell.detailTextLabel.text = dictionary[PCSPaymentTransactionDetailsLabelsContentLength];
				break;
		}
	} else if ([section.name isEqualToString:PCSPaymentTransactionDetailsOriginalTransaction]) {
		NSDictionary *dictionary = transactions[indexPath.row];

		switch (indexPath.row) {
			case PaymentTransactionDetailsSectionInformation:
				cell.textLabel.text = PCSPaymentTransactionDetailsLabelsTransactionIdentifier;
				cell.detailTextLabel.text = dictionary[PCSPaymentTransactionDetailsLabelsTransactionIdentifier];
				break;
			case PaymentTransactionDetailsSectionDownload:
				cell.textLabel.text = PCSPaymentTransactionDetailsLabelsTransactionDate;
				cell.detailTextLabel.text = dictionary[PCSPaymentTransactionDetailsLabelsTransactionDate];
				break;
		}
	} else {
		cell.textLabel.text = transactions.firstObject;
	}
}

@end

File: Swift/Common/PaymentTransactionDetails.swift
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
A BaseViewController subclass that provides details about a purchase. The purchase contains the product identifier, transaction identifier, and
transaction date for a regular purchase. It includes the content identifier, content version, and content length for a hosted product. It
contains the original transaction's identifier and date for a restored purchase.
*/

import UIKit
import StoreKit

class PaymentTransactionDetails: BaseViewController {
	// MARK: - Types

	fileprivate struct CellIdentifiers {
		static let basic = "basic"
		static let custom = "custom"
	}

	// MARK: - Properties

	var tableViewCellLabels: [SectionType: [String]] {
		let downloads = [DownloadContentLabels.contentIdentifier, DownloadContentLabels.contentVersion, DownloadContentLabels.contentLength]
		let originalTransactions = [DownloadContentLabels.transactionIdentifier, DownloadContentLabels.transactionDate]

		return [.download: downloads, .originalTransaction: originalTransactions]
	}

	// MARK: - UITable​View​Data​Source

	override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
		let section = data[indexPath.section]

		if section.type == .download || section.type == .originalTransaction {
			return tableView.dequeueReusableCell(withIdentifier: CellIdentifiers.custom, for: indexPath)
		} else {
			return tableView.dequeueReusableCell(withIdentifier: CellIdentifiers.basic, for: indexPath)
		}
	}

	// MARK: - UITableViewDelegate

	override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
		let section = data[indexPath.section]

		if section.type == .download || section.type == .originalTransaction {
			let transactions = section.elements
			guard let dictionary = transactions[indexPath.row] as? [String: String] else { return }

			let items = tableViewCellLabels[section.type]
			guard let label = items?[indexPath.row] else { fatalError("\(Messages.unknownDetail) \(indexPath.row)).") }

			cell.textLabel!.text = label
			cell.detailTextLabel!.text = dictionary[label]

		} else {
			guard let details = section.elements as? [String] else { return }
			cell.textLabel!.text = details.first
		}
	}
}

